src/librustc_mir/build/scope.rs

```diff
@@ -828,6 +879,27 @@ fn build_diverge_scope<'a, 'gcx, 'tcx>(tcx: TyCtxt<'a, 'gcx, 'tcx>,
         };
     }
 
+    let repatching;
+    if scope.diverge_path.terminator_emitted {
```

Comment:
The `EndRegion` block is preallocated, it is the `block` field on the `DivergePathState`. When `build_diverge_scope` is called, it will make that `block` end with a terminator that jumps to the drops that are built by `build_diverge_scope`.

For the purposes of instrumentation, I wanted the debug info lines to indicate whether we only do this patching once, or if we repatch the same block many times. (It turns out that we do repatch the same block many times, at least in some cases.)

And, since the `fn terminate` method requires that the block not have any terminator previously installed, we need to first install `None` to sidestep that check.

So... its a big hack, but I'm not sure more comments will help the situation.